### DISCUSSION TITLE
fcitx5 inside a Python venv

### DISCUSSION THREAD

**Discussion by ubitux** [fcitx/fcitx5#873] [Q&A] [answered]
fcitx5 cannot be triggered within a Python venv (because PySide6, the Qt Python bindings, deploys its own build of the libraries, I assume).

To reproduce, here is a simple editable Qt text widget:
```python
import sys

from PySide6.QtWidgets import QApplication, QTextEdit

app = QApplication(sys.argv)
widget = QTextEdit("Edit me")
widget.show()
app.exec()
```

Executing this Python script on a system where fcitx5 is running works just fine. On the other hand, if run within a virtual env, I'm unable to trigger fcitx:

```shell
python -m venv venv
. venv/bin/activate
pip install pyside6
python test.py
```

What are the solutions or workarounds available at this point? Would it be possible to have a fcitx pip module we could install within the venv to make the communication outside or something?

**Comment by wengxt** [comment, marked as answer]
Try set QT_IM_MODULE=ibus before launch your pyqt application. I believe in the virtual env you're not using system qt, thus you will not have fcitx5-qt installed. Note this may not work well under wayland.

If you're using KDE wayland, you can follow this https://fcitx-im.org/wiki/Using_Fcitx_5_on_Wayland#KDE_Plasma instead to make qt use its wayland input method protocol to talk to fcitx.

**Comment by ubitux** [comment]
> Try set QT_IM_MODULE=ibus before launch your pyqt application.

That works (I had to install ibus though obviously), thank you. How come this seems to still connect to fcitx5, what's the relationship here?

> I believe in the virtual env you're not using system qt, thus you will not have fcitx5-qt installed. Note this may not work well under wayland.

So I'm assuming that works because in the venv there is `lib/python3.11/site-packages/PySide6/Qt/plugins/platforminputcontexts/libibusplatforminputcontextplugin.so` deployed by the pip package `pyside6`?

> If you're using KDE wayland, you can follow this https://fcitx-im.org/wiki/Using_Fcitx_5_on_Wayland#KDE_Plasma instead to make qt use its wayland input method protocol to talk to fcitx.

Thanks, I'm using i3 (X11) though.

**Reply by wengxt** [reply]
No, you don't need to install ibus, that will make you use the qt's bundle ibus im module, and fcitx has code to simulate it.

https://github.com/qt/qtbase/blob/dev/src/plugins/platforminputcontexts/ibus/qibusplatforminputcontext.cpp
https://github.com/fcitx/fcitx5/blob/master/src/frontend/ibusfrontend/ibusfrontend.cpp

**Reply by ubitux** [reply]
Nothing happens if I don't have ibus installed (edit: even outside the venv) 🤷

**Reply by wengxt** [reply]
@ubitux I think you're right after I read the code, but I think there's a way around: 
https://github.com/qt/qtbase/blob/4f1bb8ee40f2245375ee2afef0e45cc766fcd2d4/src/plugins/platforminputcontexts/ibus/qibusplatforminputcontext.cpp#L620
https://github.com/qt/qtbase/blob/4f1bb8ee40f2245375ee2afef0e45cc766fcd2d4/src/plugins/platforminputcontexts/ibus/qibusplatforminputcontext.cpp#L606C68-L606C83

Try set
```
IBUS_USE_PORTAL=1
QT_IM_MODULE=ibus
```

In this way you should not need to have ibus installed.

**Reply by ubitux** [reply]
Awesome, this works just fine without ibus, thank you.

Any downside using this method over the "native" fcitx5?

**Reply by wengxt** [reply]
If you didn't notice any, then no.

There are some issues that I could list, but you may not use in a setup that could notice or care about them
1. on hidpi scaled system, the popup window position could be wrong
2. on wayland, the popup window position could be wrong
3. the preedit style color doesn't follow system
4. application name can not be sent to fcitx, so functionality that related to application name detection doesn't work. (share input state set to program)
5. doesn't handle input method restart well
6. different window of same qt application may only use same input method?

**Comment by ubitux** [comment]
Thank you for the workaround, your answers have been very helpful. 

I'm assuming similar issues would be raised with GTK, and I don't know if there is a possibility that such scenario could work out of the box at some point, but for now I'm closing the discussion as resolved since it solves my issue. Thanks.

**Reply by wengxt** [reply]
gtk you may use GTK_IM_MODULE=xim ,but xim is a little bit more problematic

in terms of shipping your application, you may consider flatpak which all main stream runtime has fcitx im module bundled

**Reply by ubitux** [reply]
Wouldn't it be possible to bundle fcitx as a module on pypi?

**Reply by wengxt** [reply]
That I don’t know. since it’s need to be installed as a purely cxx plugin and have nothing to do with python
